### ISSUE TITLE
Seem to have an issue downloading this video, not a crash or anything, it's as if it's simply skipped.

### ISSUE THREAD

**Issue by Montegro** [Tyrrrz/YoutubeExplode#219]
Maybe it's just my program, but I've downloaded a large amount of videos to archive them (some just a day before they were taken down) but this problem seems a little weird and I can't seem to find a way around it. Any help?
Video: https://www.youtube.com/watch?v=6RDA-YBzYY8

**Comment by Tyrrrz** [comment]
What do you mean by skipped? A download can't silently fail unless you somehow swallow an exception.

**Comment by Montegro** [comment]
I mean it gets skipped by, the issue happens after a SaveFileDialog comes up and fetches the location to store the file in the little segment of my code for it is here...
`
                            SFDFilename = SFD.FileName;
                            try
                            {
                                await CLIENT.DownloadMediaStreamAsync(STREAMINFO, $"{SFDFilename}", progressHandler, cts.Token);

                                if (IsSoundEnabled == true)
                                    System.Media.SystemSounds.Asterisk.Play();
                            }
                            catch (YoutubeExplode.Exceptions.VideoUnavailableException e)
                            {
                                DebugInfoUpdate(e.ToString());
                            }
                            catch (YoutubeExplode.Exceptions.VideoUnplayableException e)
                            {
                                DebugInfoUpdate(e.ToString());
                            }
                            catch
                            {
                                if (cancelled == true)
                                {
                                    VideoInfo.Text = "Opperation Cancelled.";
                                    DebugInfoUpdate(VideoInfo.Text);
                                }
                                else
                                {
                                    VideoInfo.Text = "An Error Has Prevented This Video From Being Saved.";
                                    DebugInfoUpdate(VideoInfo.Text);
                                }
                            }`
And the message it chooses to throw isn't caught by either of the YouTubeExplode Exceptions, it ends up throwing "An Error Has Prevented This Video From Being Saved.", which if an error would throw makes sense it wouldn't be the Unavailable or Unplayable exceptions. I've commented out my general catch and the crash error it throws ends up being an Inner Exception `IOException: Unable to read data from the transport connection: The connection was closed.` Any ideas?

**Comment by Tyrrrz** [comment]
Oh. I've seen that issue a few times but it was sporadic. Is it consistently reproducible?

**Comment by Montegro** [comment]
Seemingly I can only get it with that one video link. I'll keep an eye out to see if I can't find any others though.

**Comment by Tyrrrz** [comment]
Is it consistent on that video?

**Comment by Montegro** [comment]
Yes, I've tried at least twenty times, thirty if you include editing out catch statements to get that message, and about three times since then.

**Comment by Tyrrrz** [comment]
Yeah that sounds consistent enough. I usually would get it once in about 50 attempts. Can you check what is the `Itag` of the stream that causes these issues?

**Comment by Tyrrrz** [comment]
As annoying as it may be, the best we can do is retry, the underlying issue is caused by YouTube.

**Comment by Montegro** [comment]
I gotcha, how do I go about fetching the itag?

**Comment by Tyrrrz** [comment]
it's on `MediaStreamInfo.Itag` property

**Comment by Montegro** [comment]
Gotcha, I found it, sorry I was about to update it before you responded to say I found it. I'll update you in a moment with the itag.

**Comment by Montegro** [comment]
Here is a screengrab of my debug control window. Underneath the infofetch calls, the first number is the itag, under that is the size, and it bounces between the two of those till it hits the downloading call.
![untitled](https://user-images.githubusercontent.com/1057035/53505159-50a9ac80-3a81-11e9-8331-901698f2aea7.png)

**Comment by Tyrrrz** [comment]
I don't think there's any correlation between itags and how often they fail, you just need to use whichever fits your needs best (quality/size/etc).

**Comment by Tyrrrz** [comment]
I'll try to look into this when I have time

**Comment by Montegro** [comment]
Alright, thank you very much

**Comment by Montegro** [comment]
Could this be related to how YouTube tags videos that are unsuitable for advertisers, I'm talking the sorta tag they would give to a video containing mature content. I ran into two more instances of the same error on two more videos.
https://www.youtube.com/watch?v=L9hwGZFPSmw&list=PL34C1F26D03F5F9B8&index=28&t=0s
https://www.youtube.com/watch?v=bVzymV7P8Vs&list=PL34C1F26D03F5F9B8&t=0s&index=49

**Comment by Tyrrrz** [comment]
That should'be connected

**Comment by Tyrrrz** [comment]
I finally got some time to look into it. It seems like it's a bug in YouTube, one of the streams is just responding with `Connection: close` and not providing the content. It doesn't look identifiable ahead of time, so it seems the best way to handle it is just to `try/catch`.

**Comment by Montegro** [comment]
Is there a way to try/catch you'd recommend that could switch it to a different stream?

**Comment by Tyrrrz** [comment]
Depending on what your preference is. If you want the highest quality, I'd do something like
`mediaStreamInfoSet.Video.OrderByDescending(s => s.VideoQuality).ThenByDescending(s => s.Framerate)` and then try to download streams from that enumerable, until it succeeds.

**Comment by derekantrican** [comment]
I'm seeing this as well. Specifically this youtube video: https://www.youtube.com/watch?v=1PGm8LslEb4

Seems that the problem is caused when using `MediaStreamInfo streamInfo = streamInfoSet.Muxed.WithHighestVideoQuality()` (as suggested in the README https://github.com/Tyrrrz/YoutubeExplode#download-video )

If I use `MediaStreamInfo streamInfo = streamInfoSet.Video.OrderByDescending(s => s.VideoQuality).First();` or `MediaStreamInfo streamInfo = streamInfoSet.Video.OrderByDescending(s => s.VideoQuality).ThenByDescending(s => s.Framerate).First();` like you suggested, there is no problem.

Would love to see this fixed in a future version

**Comment by Tyrrrz** [comment]
As I said above, this can't really be "fixed" as it's a problem with YouTube. The best we can do is probe each stream to ensure that all of them return content. Currently only muxed streams that lack content length metadata are being probed with HEAD requests, which is about 0-2 streams per stream info set. Even then, the faulty streams respond the same way to HEAD requests as any other streams do. The only way to "check" if a stream is faulty is to actually open a connection and try to read at least the first few bytes. This will induce a big performance hit if we were to do it on all streams in the set, which is on average around 30 streams.

The best way to handle this issue is to try/catch with fallback and wait until YouTube (hopefully) fixes it.

**Comment by Tyrrrz** [comment]
Glad it helped 😊

**Comment by Tyrrrz** [comment]
I improved the readme to include more examples

**Comment by derekantrican** [comment]
I know this is a closed issue, but I am getting zero qualities for pretty much any video now. This pretty much renders YouTubeExplode useless for me. Here's what I'm doing (based on earlier suggestions):

````
var streamInfoSet = await client.GetVideoMediaStreamInfosAsync(youTubeVideoId);
List<VideoStreamInfo> videoQualities = streamInfoSet.Video.OrderByDescending(s => s.VideoQuality).ToList();
List<MuxedStreamInfo> muxedQualities = streamInfoSet.Muxed.OrderByDescending(s => s.VideoQuality).ToList();
````
(These lists are always empty for videos that have been uploaded recently)

Any suggestions?
